safenode/src/node/api.rs

```diff
@@ -113,118 +112,123 @@ impl Node {
-            let error = ProtocolError::SignedSrcTxHashDoesNotMatchProvidedSrcTxHash {
-                signed_src_tx_hash,
-                provided_src_tx_hash,
-            };
-            self.send_response(
-                Response::Cmd(CmdResponse::Spend(Err(error))),
-                response_channel,
-            )
-            .await;
-            return Ok(());
+    async fn handle_query(&mut self, query: Query) -> QueryResponse {
+        match query {
+            Query::Register(query) => self.registers.read(&query, User::Anyone).await,
+            Query::GetChunk(address) => {
+                let resp = self.chunks.get(&address).await;
+                QueryResponse::GetChunk(resp)
+            }
+            Query::Spend(query) => {
+                match query {
+                    SpendQuery::GetFees { dbc_id, priority } => {
+                        // The client is asking for the fee to spend a specific dbc, and including the id of that dbc.
+                        // The required fee content is encrypted to that dbc id, and so only the holder of the dbc secret
+                        // key can unlock the contents.
+                        let required_fee = self.transfers.get_required_fee(dbc_id, priority);
+                        QueryResponse::GetFees(Ok(required_fee))
+                    }
+                    SpendQuery::GetDbcSpend(address) => {
+                        let res = self
+                            .transfers
+                            .get(address)
+                            .await
+                            .map_err(ProtocolError::Transfers);
+                        QueryResponse::GetDbcSpend(res)
+                    }
+                }
+            }
         }
+    }
 
-        // First we need to validate the parents of the spend.
-        // This also ensures that all parent's dst tx's are the same as the src tx of this spend.
-        match self
-            .validate_spend_parents(signed_spend.as_ref(), source_tx.as_ref())
-            .await
-        {
-            Ok(()) => (),
-            Err(super::Error::Protocol(error)) => {
-                // We drop spend attempts with invalid parents,
-                // and return an error to the client.
-                self.send_response(
-                    Response::Cmd(CmdResponse::Spend(Err(error))),
-                    response_channel,
-                )
-                .await;
-                return Ok(());
+    async fn handle_cmd(&mut self, cmd: Cmd) -> CmdResponse {
+        match cmd {
+            Cmd::StoreChunk(chunk) => {
+                let resp = self.chunks.store(&chunk).await;
+                CmdResponse::StoreChunk(resp)
             }
-            // This should be unreachable, as we only return protocol errors.
-            other => other?,
-        };
-
-        let response = match self
-            .storage
-            .write(&Cmd::Dbc {
+            Cmd::Register(cmd) => {
```

Comment:
Should these commands perhaps be called `WriteRegister` and `SpendDbc`?

Reply:
When we have many levels, we have been categorizing them at the higher level, providing the cmd description in the lower level. 

RegisterCmd has Create + Edit.

What category of cmd? Register. What type of register cmd? Edit.

Dbc has only one variant - so far - and could be SpendDbc there doesn't seem to be more variants to be added there.

Reply:
(The `Cmd::Dbc` variant was renamed to `SpendDbc`.)